common/blast-query-parser/src/main/kotlin/mblast/query/pipe/SequenceFileWriter.kt

```diff
@@ -0,0 +1,219 @@
+   *
+   * The returned result will also contain a list of sub-query files for each
+   * individual sequence **if and only if** the input query contained more than
+   * one sequence.
+   *
+   * If the input stream contains only one sequence, the returned result
+   * sub-query list will be empty as the overall query file is the only result.
+   */
+  override fun run() {
+    try {
+      super.run()
+    } catch (e: Throwable) {
+      cleanUp()
+      throw e
+    } finally {
+      close()
+    }
+  }
+
+  fun getResult() = SequenceFileWriteResult(overallOutputFile, childOutputFiles)
+
+  override fun handleChar(c: Int) {
+    if (c > -1) {
+      overallOutputStream.write(c)
+      currentChildStream?.write(c)
+    }
+  }
+
+  override fun nextCharacter() = stream.read()
+
+  override fun onSequenceStart() {
+    when (sequenceCount) {
+      // If this was called before the first sequence was hit, then something
+      // has gone wrong in the underlying state machine.
+      0 -> throw IllegalStateException()
+
+      // If this is the first sequence, then there is nothing to do.
+      // If there is only one sequence then we only need the overall query file.
+      1 -> {}
+
+      // If this is the second sequence, we know that we are going to need
+      // sub-query files for the individual sequences.
+      2 -> {
+        // Since there are multiple sequences in this query, we can't just have
+        // the one file represent the whole thing, copy over the first sequence
+        // to its own file before continuing to write to the overall query file.
+        copyOverallFirstChild()
+
+        // Set up a file for the coming sequence.
+        newChild()
+      }
+
+      // Else we are on the third sequence or later and we have already
+      // established that we need sub-query files.
+      // Set up the next one.
+      else -> newChild()
+    }
+  }
+
+  override fun onQueryComplete() {
+    overallOutputStream.flush()
+    overallOutputStream.close()
+    currentChildStream?.flush()
+    currentChildStream?.close()
+  }
+
+  override fun close() {
+    stream.close()
+    overallOutputStream.close()
+    currentChildStream?.close()
+  }
+
+  /**
+   * Closes the current child file and stream and opens a new one.
+   */
+  private fun newChild() {
+    currentChildStream?.also {
+      it.flush()
+      it.close()
+    }
+
+    // + 1 because index 0 is reserved for the overall file.
+    currentChildFile   = makeFile(childOutputFiles.size + 1).also {
+      currentChildStream = it.outputStream().buffered()
+      childOutputFiles.add(it)
+    }
+  }
+
+  /**
+   * Copies the contents of the overall output file to a new child file.
+   */
+  private fun copyOverallFirstChild() {
+    // Create a new child file
+    newChild()
+
+    // Copy over the contents of our overall file
+    overallOutputStream.flush()
+    overallOutputFile.inputStream().use { it.transferTo(currentChildStream) }
```

Comment:
Not so relevant to this PR, but I don't really understand the structure of these BLAST parent files and child files. This is probably something I should read up on at some point.

Reply:
For multi-sequence queries, there is an overall/full query file which contains the entire user input, then there are child files which contain single individual sequences pulled out of the full query.

So for example, given this input:
```
> sequence 1
catcatcat
> sequence 2
atatataata
```
there would be 3 total files generated:
```
query-0.txt  // The full query text
query-1.txt  // only sequence 1
query-2.txt  // only sequence 2
```

For single  sequence queries, there is only a single file as the overall query _is_ the first single sequence.